sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AbstractSqlParser.scala

```diff
@@ -30,50 +32,74 @@ abstract class AbstractSqlParser extends AbstractParser with ParserInterface {
   override def astBuilder: AstBuilder
 
   /** Creates Expression for a given SQL string. */
-  override def parseExpression(sqlText: String): Expression = parse(sqlText) { parser =>
-    val ctx = parser.singleExpression()
-    withOrigin(ctx, Some(sqlText)) {
-      astBuilder.visitSingleExpression(ctx)
+  override def parseExpression(sqlText: String): Expression =
+    parse(sqlText) { parser =>
+      val ctx = parser.singleExpression()
+      withOrigin[Expression](ctx, Some(sqlText)) {
+        withErrorHandling(ctx) {
+          astBuilder.visitSingleExpression(ctx)
+        }
+      }
     }
-  }
 
   /** Creates TableIdentifier for a given SQL string. */
-  override def parseTableIdentifier(sqlText: String): TableIdentifier = parse(sqlText) { parser =>
-    astBuilder.visitSingleTableIdentifier(parser.singleTableIdentifier())
```

Comment:
Looking at methods in this file, I think all of them should do `withOrigin(ctx, Some(sqlText))` but some are missing. Now it's a good chance to fix it, by doing it in the new `withErrorHandling` function.

Reply:
oh nvm, some do not have parser context